ffmpeg/dag.py

```diff
@@ -69,31 +70,38 @@ def incoming_edge_map(self):
         raise NotImplementedError()
 
 
-DagEdge = namedtuple('DagEdge', ['downstream_node', 'downstream_label', 'upstream_node', 'upstream_label'])
+DagEdge = namedtuple('DagEdge', ['downstream_node', 'downstream_label', 'upstream_node', 'upstream_label', 'upstream_selector'])
 
 
 def get_incoming_edges(downstream_node, incoming_edge_map):
     edges = []
-    for downstream_label, (upstream_node, upstream_label) in list(incoming_edge_map.items()):
-        edges += [DagEdge(downstream_node, downstream_label, upstream_node, upstream_label)]
+    # downstream_label, (upstream_node, upstream_label) in [(i[0], i[1][:2]) for i in self.incoming_edge_map.items()]
+    for downstream_label, upstream_info in [(i[0], i[1]) for i in incoming_edge_map.items()]:
+        upstream_node, upstream_label = upstream_info[:2]
+        upstream_selector = None if len(upstream_info) < 3 else upstream_info[2]
+        edges += [DagEdge(downstream_node, downstream_label, upstream_node, upstream_label, upstream_selector)]
     return edges
 
 
 def get_outgoing_edges(upstream_node, outgoing_edge_map):
     edges = []
     for upstream_label, downstream_infos in list(outgoing_edge_map.items()):
-        for (downstream_node, downstream_label) in downstream_infos:
-            edges += [DagEdge(downstream_node, downstream_label, upstream_node, upstream_label)]
+        for downstream_info in downstream_infos:
+            downstream_node, downstream_label = downstream_info[:2]
+            downstream_selector = None if len(downstream_info) < 3 else downstream_info[2]
+            edges += [DagEdge(downstream_node, downstream_label, upstream_node, upstream_label, downstream_selector)]
     return edges
 
 
 class KwargReprNode(DagNode):
     """A DagNode that can be represented as a set of args+kwargs.
     """
+
     @property
     def __upstream_hashes(self):
         hashes = []
-        for downstream_label, (upstream_node, upstream_label) in list(self.incoming_edge_map.items()):
+        # This is needed to allow extra stuff in the incoming_edge_map's value tuples
+        for downstream_label, (upstream_node, upstream_label) in [(i[0], i[1][:2]) for i in self.incoming_edge_map.items()]:
```

Comment:
What are `i[0]`, `i[1]`, etc.? Should make this code more clear and split it out into something easier to understand. Way too much happening in this line.

Reply:
Done 14b64706